enclave/src/polkadex_gateway.rs

```diff
@@ -0,0 +1,137 @@
+use log::*;
+use polkadex_sgx_primitives::AccountId;
+use polkadex_sgx_primitives::types::{Order, OrderSide, OrderType, OrderUUID, TradeEvent};
+use sgx_types::{sgx_status_t, SgxResult};
+
+use crate::polkadex;
+use crate::polkadex_balance_storage;
+use crate::polkadex_orderbook_storage;
+
+/// Place order function does the following
+/// 1. authenticate
+/// 2. mutate balances (reserve amount offered in order)
+/// 3. store_order (async)
+/// 4. send order to OpenFinex API
+/// 5. report OpenFinex API result to sender
+pub fn place_order(main_account: AccountId, proxy_acc: Option<AccountId>, order: Order) -> SgxResult<()> {
+    // Authentication
+    authenticate_user(main_account.clone(), proxy_acc)?;
+    // Mutate Balances
+    match order.order_type {
+        OrderType::LIMIT => {
+            if let Some(price) = order.price {
+                match order.side {
+                    OrderSide::BID => {
+                        // TODO: Is it safe to use saturating_mul here?
```

Comment:
saturating_mul will make sure enclave doesn't panic but is it okay? Blockchain represents 1 token as 10^18 in u128, so if someone places 1 token as price and 1 token as qty, then 10^18 x 10^18 happens. How can we handle this?

Reply:
@brenzi, this one as well..

Reply:
Basically, multiplication must happen with u256 and we then take the MSB u128 as a result (actually, treat it like a fixpoint. The number of decimals will define the bitshift from u256 to u128)
In any case we have to avoid panic! su sat_mul should be used and properly error-handled

Reply:
we do lose precision here, that's clear. but it will be dust at worst

Reply:
@brenzi, is there any library to do this? or we have to implement it?

Reply:
I am using this here https://github.com/encointer/substrate-fixed